### DISCUSSION TITLE
Weird error with umqtt script and homeassistant

### DISCUSSION THREAD

**Discussion by liamriley1** [micropython/micropython#13465] [RP2040 / Pico] [answered]
Hi!

I've been trying to integrate my garage door into homeassistant with a raspberry pi pico w and MQTT (mosquitto server). I'm currently just trying to get the MQTT autodiscovery part to work fully, and I had this fully working section:

```
## Send discovery message to homeasstant so that it sees this device! ##
def send_ha_discovery():
    discovery_topic = b"home/cover/garage-door/config"

    discovery_payload = {
                    "~":"home/garage/garage-door",
                    "name":"garage-door",
                    "uniq_id": "garage-door",
                    "cmd_t":"~/ctrl",
                    "stat_t":"~/status",
                    "device_class": "garage"
                    }
    client.publish(discovery_topic, json.dumps(discovery_payload))

send_ha_discovery()


### Send our initial status as closed ###
client.publish(b"home/garage/garage-door/status", b"closed")

while True:
    ### Check for new messages from MQTT broker ###
    client.wait_msg()

```
I then updated the device_payload section with the "device" key to add more information in homeasisstant:

 ```
   discovery_payload = {
                    "~":"home/garage/garage-door",
                    "name":"garage-door",
                    "uniq_id": "garage-door",
                    "cmd_t":"~/ctrl",
                    "stat_t":"~/status",
                    "device_class": "garage",
                    "device":
                        {"manufacturer": "Hörmann",
                         "model": "SupraMatic E2",
                         "suggested_area": "garage"}
                    }
```

Now, when I tried to run everything, I got an` OSError: [Errno 104] ECONNRESET` error in the console. When I checked the logs for mosquitto, I saw this: `Client picow-test disconnected due to malformed packet.`

I can't really see how this addition could cause this error?

To add more confusion, I can get the code with the new `device_payload` working again, but I have to comment out an unrelated MQTT publish statement, even though the python erorr message comes at the line with the `client.wait_msg() `part, where the pico tries to poll the MQTT server! I'm really at a loss as to how the extra addition to the json data can cause a malformed packet, and how a publish statement on another line seems to then break the MQTT connection when I try and check for messages!

Any help from people with experience with MQTT would be really appreciated, thanks!

**Comment by peterhinch** [comment]
I can't explain why the error is occurring, but the fact that it appears in `.wait_msg` seems clear. The sequence is as follows. The publication occurs, and the transfer proceeds to completion - `.publish` throws no exception. The broker then decides it doesn't like what it got and closes the connection. The first the application knows about this is when it next tries to access the broker.

In my experience mosquitto is rock-solid reliable. However the official MQTT clients have limitations. You might like to look at [mqtt_as](https://github.com/peterhinch/micropython-mqtt/blob/master/mqtt_as/README.md) - the doc explains some of the problems with the official modules.

**Reply by liamriley1** [reply, marked as answer]
Thanks for your help! It seemed like the error was to do with the end of the packet being cut off for some reason, but I managed to solve it in the end by just randomly moving things round until it started working again :| definitely not the best way to fix stuff but it works now so I'm happy :)
